python/llm/test/inference_gpu/test_transformers_api_decoder.py

```diff
@@ -0,0 +1,120 @@
+            layer_tensor = self.layer_outputs.pop()
+            model.to('cpu')
+ 
+            opt_model = Model.from_pretrained(model_path,
+                                            load_in_4bit=True,
+                                            optimize_model=True,
+                                            trust_remote_code=True)
+            opt_model = opt_model.to(device)
+ 
+ 
+            def replace_forward_hook(module, input, output, layer_name):
+                output = self.pre_layer_outputs[0]
+                return output
+ 
+            for layer_name, layer_module in opt_model.named_modules():
+                if layer_name == layer_before_test_layer:
+                    layer_module.register_forward_hook(
+                        lambda module, input, output, layer_name=layer_name: replace_forward_hook(module, input,
+                                                                                                output, layer_name))
+                if layer_name == test_layer:
+                    layer_module.register_forward_hook(
+                        lambda module, input, output, layer_name=layer_name: forward_hook(module, input,
+                                                                                        output, layer_name))
+            logits_optimized_model = (opt_model(input_ids)).logits
+            # the list `layer_output` has only one element.
+            opt_layer_tensor = self.layer_outputs[0]
+            opt_model.to('cpu')
+ 
+ 
+            test_layer_output_diff = []
+            for i, (t1, t2) in enumerate(zip(layer_tensor, opt_layer_tensor)):
+                if t1 is not None and t2 is not None:
+                    if isinstance(t1, torch.Tensor) and isinstance(t2, torch.Tensor):
+                        test_layer_output_diff.append(t1 - t2)
+                    else:
+                        # 'past_key_value'is of type tuple as default.
+                        for i, (t3, t4) in enumerate(zip(t1, t2)):
+                            test_layer_output_diff.append(t3 - t4)
+ 
+            max_diff_tensor = [torch.max(item).item() for item in test_layer_output_diff]
+            print(max_diff_tensor)
+           
+            assert all(max_diff <= lower_bound for max_diff in max_diff_tensor)
+   
+    @pytest.mark.parametrize('Name, Model, Tokenizer, model_path',TEST_MODEL_LIST)
+    def test_dynamic_functions(self, Name, Model, Tokenizer, model_path):
+        if Name == "Llama2-7B":
+            self.Llama2_7B_gpu_model(Name, Model, Tokenizer, model_path)
+
+
+    def Llama2_7B_gpu_model(self, Name, Model, Tokenizer, model_path):
+        # currently only compare the output of the last Decoder layer.
+        layer_before_Decoder = "model.layers.30"
+        Decoder_layer = "model.layers.31"
+        lower_bound = 5e-2
+        self.run_optimize_gpu_model(Name, Model, Tokenizer, model_path, Decoder_layer, layer_before_Decoder, lower_bound)
```

Comment:
https://github.com/intel-analytics/BigDL/pull/9932 decoder is for fp16 and no need to be tested? for int4 it is covered in mlp.